## Market Analysis - I

<img src="images.jpg" width="400" align="center">

This is the first of two sections on market analysis on an Online Retail <a href= https://www.kaggle.com/vijayuv/onlineretail>dataset</a>

#### Highlights:

#### Part I
* <a href='#1.0_destination'>Select a target Metric</a>
* <a href='#1.1_destination'>Monthly Growth Rate</a>
* <a href='#1.2_destination'>Monthly Order Count</a>

#### Part II
* <a href='#2_destination'>New Customer Ratio</a>
* <a href='#3_destination'>Retention Rate</a>




## Part I

With the following Metrics: 

* Customer ID
* Unit Price
* Quantity
* Invoice Date

We can build our target metric based on this equation

**Revenue = Active Customer Count * Order Count * Average Revenue per Order**

Also we will obtain:

**New Customer Ratio**:  New cusomer/Existing customers is a good indicator of if we are losing our existing customers or unable to attract new ones

In [ ]:
# import libraries
from __future__ import division
from datetime import datetime, timedelta
import pandas as pd
%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns


import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

In [4]:
#Load Data
tx_data = pd.read_csv('OnlineRetail.csv',encoding='ISO-8859-1')

tx_data.head(10)

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
5    536365     22752         SET 7 BABUSHKA NESTING BOXES         2   
6    536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER         6   
7    536366     22633               HAND WARMER UNION JACK         6   
8    536366     22632            HAND WARMER RED POLKA DOT         6   
9    536367     84879        ASSORTED COLOUR BIRD ORNAMENT        32   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom  
5  12/1/2010 8:26       7.65     17850.0  United Kingdom  
6  12/1/2010 8:26       4.25     17850.0  United Kingdom  
7  12/1/2010 8:28       1.85     17850.0  United Kingdom  
8  12/1/2010 8:28       1.85     17850.0  United Kingdom  
9  12/1/2010 8:34       1.69     13047.0  United Kingdom

In [19]:
#shape of Data
tx_data.shape

(541909, 10)

In [20]:
tx_data.describe()

Quantity      UnitPrice     CustomerID  InvoiceYearMonth  \
count  541909.000000  541909.000000  406829.000000     541909.000000   
mean        9.552250       4.611114   15287.690570     201099.713989   
std       218.081158      96.759853    1713.600303         25.788703   
min    -80995.000000  -11062.060000   12346.000000     201012.000000   
25%         1.000000       1.250000   13953.000000     201103.000000   
50%         3.000000       2.080000   15152.000000     201107.000000   
75%        10.000000       4.130000   16791.000000     201110.000000   
max     80995.000000   38970.000000   18287.000000     201112.000000   

             Revenue  
count  541909.000000  
mean       17.987795  
std       378.810824  
min   -168469.600000  
25%         3.400000  
50%         9.750000  
75%        17.400000  
max    168469.600000

###  <a id='1.0_destination'>Target Metric</a>


In [5]:
#converting the type of Invoice Date Field from string to datetime.
tx_data['InvoiceDate'] = pd.to_datetime(tx_data['InvoiceDate'])

In [9]:
#creating YearMonth field for the ease of reporting and visualization
tx_data['InvoiceYearMonth'] = tx_data['InvoiceDate'].map(lambda date: 100*date.year + date.month)

In [7]:
#calculate Revenue for each row and create a new dataframe with YearMonth - Revenue columns
tx_data['Revenue'] = tx_data['UnitPrice'] * tx_data['Quantity']
tx_revenue = tx_data.groupby(['InvoiceYearMonth'])['Revenue'].sum().reset_index()
tx_revenue

InvoiceYearMonth      Revenue
0             201012   748957.020
1             201101   560000.260
2             201102   498062.650
3             201103   683267.080
4             201104   493207.121
5             201105   723333.510
6             201106   691123.120
7             201107   681300.111
8             201108   682680.510
9             201109  1019687.622
10            201110  1070704.670
11            201111  1461756.250
12            201112   433668.010

In [12]:
#Visualizing Revenue growth over these period 
plot_data = [
    go.Scatter(
        x=tx_revenue['InvoiceYearMonth'],
        y=tx_revenue['Revenue']
    )
]

plot_layout = go.Layout(
        xaxis={"type": "category"},
        title='Montly Revenue'
    )
fig = go.Figure(data=plot_data, layout=plot_layout)
fig.show()

### <a id='1.1_destination'>Monthly growth rate</a>

In [15]:

#using pct_change() function to see monthly percentage change
tx_revenue['MonthlyGrowth'] = tx_revenue['Revenue'].pct_change()

#showing first 5 rows
tx_revenue



InvoiceYearMonth      Revenue  MonthlyGrowth
0             201012   748957.020            NaN
1             201101   560000.260      -0.252293
2             201102   498062.650      -0.110603
3             201103   683267.080       0.371850
4             201104   493207.121      -0.278163
5             201105   723333.510       0.466592
6             201106   691123.120      -0.044530
7             201107   681300.111      -0.014213
8             201108   682680.510       0.002026
9             201109  1019687.622       0.493653
10            201110  1070704.670       0.050032
11            201111  1461756.250       0.365228
12            201112   433668.010      -0.703324

In [18]:
#visualization - line graph
plot_data = [
    go.Scatter(
        x=tx_revenue.query("InvoiceYearMonth < 201112")['InvoiceYearMonth'],
        y=tx_revenue.query("InvoiceYearMonth < 201112")['MonthlyGrowth']
    )
]

plot_layout = go.Layout(
        xaxis={"type": "category"},
        title='Monthly Growth Rate'
    )

fig = go.Figure(data=plot_data, layout=plot_layout)
fig.show()

April appears to be the lowest point in the plot, hence indicates a **point of interest**

###  Who are the Monthly Active Customers?

In [21]:
#creating a new dataframe with UK customers only
tx_uk = tx_data.query("Country=='United Kingdom'").reset_index(drop=True)

#creating monthly active customers dataframe by counting unique Customer IDs
tx_monthly_active = tx_uk.groupby('InvoiceYearMonth')['CustomerID'].nunique().reset_index()

#print the dataframe
tx_monthly_active



InvoiceYearMonth  CustomerID
0             201012         871
1             201101         684
2             201102         714
3             201103         923
4             201104         817
5             201105         985
6             201106         943
7             201107         899
8             201108         867
9             201109        1177
10            201110        1285
11            201111        1548
12            201112         617

In [23]:
#plotting monthly active customers
plot_data = [
    go.Bar(
        x=tx_monthly_active['InvoiceYearMonth'],
        y=tx_monthly_active['CustomerID'],
    )
]

plot_layout = go.Layout(
        xaxis={"type": "category"},
        title='Monthly Active Customers'
    )

fig = go.Figure(data=plot_data, layout=plot_layout)
fig.show()

In April, Monthly Active Customer number dropped to 817 from 923 (-11.5%).

### <a id='1.2_destination'>Monthly Order Count </a>
We repeat the same process as above using the Quantity field:


In [24]:
#create a new dataframe for no. of order by using quantity field
tx_monthly_sales = tx_uk.groupby('InvoiceYearMonth')['Quantity'].sum().reset_index()

#print the dataframe
tx_monthly_sales

InvoiceYearMonth  Quantity
0             201012    298101
1             201101    237381
2             201102    225641
3             201103    279843
4             201104    257666
5             201105    306452
6             201106    258522
7             201107    324129
8             201108    319804
9             201109    458490
10            201110    455612
11            201111    642281
12            201112    199907

In [30]:
#plotting monthly quantity
plot_data = [
    go.Bar(
        x=tx_monthly_sales['InvoiceYearMonth'],
        y=tx_monthly_sales['Quantity'],
    )
]

plot_layout = go.Layout(
        xaxis={"type": "category"},
        title='Monthly Total No of Order'
    )

fig = go.Figure(data=plot_data, layout=plot_layout)
fig.show()

Order Count also declined in April (279k to 257k, -8%). Obviously Active Customer Count directly affected Order Count decrease. 
Subsequently we check our Average Revenue per Order as well.

### Average Revenue per Order

To get this data, we need to calculate the average of revenue for each month:

In [28]:
# create a new dataframe for average revenue by taking the mean of it
tx_monthly_order_avg = tx_uk.groupby('InvoiceYearMonth')['Revenue'].mean().reset_index()

#print the dataframe
tx_monthly_order_avg

InvoiceYearMonth    Revenue
0             201012  16.865860
1             201101  13.614680
2             201102  16.093027
3             201103  16.716166
4             201104  15.773380
5             201105  17.713823
6             201106  16.714748
7             201107  15.723497
8             201108  17.315899
9             201109  18.931723
10            201110  16.093582
11            201111  16.312383
12            201112  16.247406

In [29]:
#plot the data
plot_data = [
    go.Bar(
        x=tx_monthly_order_avg['InvoiceYearMonth'],
        y=tx_monthly_order_avg['Revenue'],
    )
]

plot_layout = go.Layout(
        xaxis={"type": "category"},
        title='Monthly Order Average'
    )
fig = go.Figure(data=plot_data, layout=plot_layout)
fig.show()

In [ ]:
Clearly the monthly order average dropped for April (16.7 to 15.8)

## Part ii

We investigate some other important metrics namely:

**New Customer Ratio**:  New cusomer/Existing customers is a good indicator of if we are losing our existing customers or unable to attract new ones
    
**Retention Rate**: Indicates how many customers we retain over specific time window. 

### <a id='2_destination'>New Customer Ratio</a>

In [55]:
#create a dataframe contaning CustomerID and first purchase date
tx_min_purchase = tx_uk.groupby('CustomerID').InvoiceDate.min().reset_index()

tx_min_purchase.columns = ['CustomerID','MinPurchaseDate']

tx_min_purchase['MinPurchaseYearMonth'] = tx_min_purchase['MinPurchaseDate'].map(lambda date: 100*date.year + date.month)

tx_min_purchase.head(5)

CustomerID     MinPurchaseDate  MinPurchaseYearMonth
0     12346.0 2011-01-18 10:01:00                201101
1     12747.0 2010-12-05 15:38:00                201012
2     12748.0 2010-12-01 12:48:00                201012
3     12749.0 2011-05-10 15:25:00                201105
4     12820.0 2011-01-17 12:34:00                201101

In [39]:
#merge first purchase date column to our main dataframe (tx_uk)
tx_uk = pd.merge(tx_uk, tx_min_purchase, on='CustomerID')

tx_uk.head(5)

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  \
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom   
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom   
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom   
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom   
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom   

   InvoiceYearMonth  Revenue   MinPurchaseDate_x  MinPurchaseYearMonth_x  \
0            201012    15.30 2010-12-01 08:26:00                  201012   
1            201012    20.34 2010-12-01 08:26:00                  201012   
2            201012    22.00 2010-12-01 08:26:00                  201012   
3            201012    20.34 2010-12-01 08:26:00                  201012   
4            201012    20.34 2010-12-01 08:26:00                  201012   

    MinPurchaseDate_y  MinPurchaseYearMonth_y UserType     MinPurchaseDate  \
0 2010-12-01 08:26:00                  201012      New 2010-12-01 08:26:00   
1 2010-12-01 08:26:00                  201012      New 2010-12-01 08:26:00   
2 2010-12-01 08:26:00                  201012      New 2010-12-01 08:26:00   
3 2010-12-01 08:26:00                  201012      New 2010-12-01 08:26:00   
4 2010-12-01 08:26:00                  201012      New 2010-12-01 08:26:00   

   MinPurchaseYearMonth  
0                201012  
1                201012  
2                201012  
3                201012  
4                201012

In [41]:
#create a column called User Type and assign Existing 
th
tx_uk['UserType'] = 'New'

#if User's First Purchase Year Month before the selected Invoice Year Mon
tx_uk.loc[tx_uk['InvoiceYearMonth']>tx_uk['MinPurchaseYearMonth'],'UserType'] = 'Existing'

In [42]:
#calculate the Revenue per month for each user type
tx_user_type_revenue = tx_uk.groupby(['InvoiceYearMonth','UserType'])['Revenue'].sum().reset_index()

In [44]:
tx_uk.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  \
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom   
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom   
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom   
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom   
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom   

   InvoiceYearMonth  Revenue   MinPurchaseDate_x  MinPurchaseYearMonth_x  \
0            201012    15.30 2010-12-01 08:26:00                  201012   
1            201012    20.34 2010-12-01 08:26:00                  201012   
2            201012    22.00 2010-12-01 08:26:00                  201012   
3            201012    20.34 2010-12-01 08:26:00                  201012   
4            201012    20.34 2010-12-01 08:26:00                  201012   

    MinPurchaseDate_y  MinPurchaseYearMonth_y UserType     MinPurchaseDate  \
0 2010-12-01 08:26:00                  201012      New 2010-12-01 08:26:00   
1 2010-12-01 08:26:00                  201012      New 2010-12-01 08:26:00   
2 2010-12-01 08:26:00                  201012      New 2010-12-01 08:26:00   
3 2010-12-01 08:26:00                  201012      New 2010-12-01 08:26:00   
4 2010-12-01 08:26:00                  201012      New 2010-12-01 08:26:00   

   MinPurchaseYearMonth  
0                201012  
1                201012  
2                201012  
3                201012  
4                201012

In [47]:
#filtering the dates and plot the result
tx_user_type_revenue = tx_user_type_revenue.query("InvoiceYearMonth != 201012 and InvoiceYearMonth != 201112")
tx_user_type_revenue 

InvoiceYearMonth  UserType     Revenue
1             201101  Existing  195275.510
2             201101       New  156705.770
3             201102  Existing  220994.630
4             201102       New  127859.000
5             201103  Existing  296350.030
6             201103       New  160567.840
7             201104  Existing  268226.660
8             201104       New  108517.751
9             201105  Existing  434725.860
10            201105       New   90847.490
11            201106  Existing  408030.060
12            201106       New   64479.190
13            201107  Existing  407693.610
14            201107       New   53453.991
15            201108  Existing  421388.930
16            201108       New   55619.480
17            201109  Existing  640861.901
18            201109       New  135667.941
19            201110  Existing  648837.600
20            201110       New  133940.280
21            201111  Existing  838955.910
22            201111       New  117153.750

In [48]:
#Revenue of New vs Existing
plot_data = [
    go.Scatter(
        x=tx_user_type_revenue.query("UserType == 'Existing'")['InvoiceYearMonth'],
        y=tx_user_type_revenue.query("UserType == 'Existing'")['Revenue'],
        name = 'Existing'
    ),
    go.Scatter(
        x=tx_user_type_revenue.query("UserType == 'New'")['InvoiceYearMonth'],
        y=tx_user_type_revenue.query("UserType == 'New'")['Revenue'],
        name = 'New'
    )
]

plot_layout = go.Layout(
        xaxis={"type": "category"},
        title='New vs Existing'
    )
fig = go.Figure(data=plot_data, layout=plot_layout)
fig.show()

Clearly much of the Revenue comes from existing customers. Existing customers shows a positive trend and tell us that the customer base is growing but new customers have a slight negative trend.

A better view by looking at the New Customer Ratio:

In [51]:
#create a dataframe that shows new user ratio - we also need to drop NA values (first month new user ratio is 0)
tx_user_ratio = tx_uk.query("UserType == 'New'").groupby(['InvoiceYearMonth'])['CustomerID'].nunique()/tx_uk.query("UserType == 'Existing'").groupby(['InvoiceYearMonth'])['CustomerID'].nunique() 
tx_user_ratio = tx_user_ratio.reset_index()
tx_user_ratio = tx_user_ratio.dropna()
tx_user_ratio 

InvoiceYearMonth  CustomerID
1             201101    1.124224
2             201102    0.904000
3             201103    0.792233
4             201104    0.510166
5             201105    0.343793
6             201106    0.281250
7             201107    0.236589
8             201108    0.192572
9             201109    0.304878
10            201110    0.328852
11            201111    0.236422
12            201112    0.058319

In [59]:
#plot the ratio of New Against Exixting customers

plot_data = [
    go.Bar(
        x=tx_user_ratio.query("InvoiceYearMonth>201101 and InvoiceYearMonth<201112")['InvoiceYearMonth'],
        y=tx_user_ratio.query("InvoiceYearMonth>201101 and InvoiceYearMonth<201112")['CustomerID'],
    )
]

plot_layout = go.Layout(
        xaxis={"type": "category"},
        title='New Customer Ratio'
    )
fig = go.Figure(data=plot_data, layout=plot_layout)
fig.show()

The Bar chart clearly show a decline in new cutomers.

###   <a id='3_destination'>Retention Rate</a>

Retention rate should be monitored very closely because it indicates how sticky is your service and how well your product fits 

the market. For making Monthly Retention Rate visualized, we need to calculate how many customers retained from previous month.

**Monthly Retention Rate** = Retained Customers From Prev. Month/Active Customers Total

In [66]:
#identify which users are active by looking at their revenue per month
tx_user_purchase = tx_uk.groupby(['CustomerID','InvoiceYearMonth'])['Revenue'].sum().reset_index()
tx_user_purchase

CustomerID  InvoiceYearMonth  Revenue
0         12346.0            201101     0.00
1         12747.0            201012   706.27
2         12747.0            201101   303.04
3         12747.0            201103   310.78
4         12747.0            201105   771.31
...           ...               ...      ...
12325     18283.0            201110   114.65
12326     18283.0            201111   651.56
12327     18283.0            201112   208.00
12328     18287.0            201105   765.28
12329     18287.0            201110  1072.00

[12330 rows x 3 columns]

In [61]:
#create retention matrix with crosstab
tx_retention = pd.crosstab(tx_user_purchase['CustomerID'], tx_user_purchase['InvoiceYearMonth']).reset_index()

tx_retention.head()

InvoiceYearMonth  CustomerID  201012  201101  201102  201103  201104  201105  \
0                    12346.0       0       1       0       0       0       0   
1                    12747.0       1       1       0       1       0       1   
2                    12748.0       1       1       1       1       1       1   
3                    12749.0       0       0       0       0       0       1   
4                    12820.0       0       1       0       0       0       0   

InvoiceYearMonth  201106  201107  201108  201109  201110  201111  201112  
0                      0       0       0       0       0       0       0  
1                      1       0       1       0       1       1       1  
2                      1       1       1       1       1       1       1  
3                      0       0       1       0       0       1       1  
4                      0       0       0       1       1       0       1

Retention table shows us which customers are active on each month (1 stands for active).

In [62]:
#create an array of dictionary which keeps Retained & Total User count for each month
months = tx_retention.columns[2:]
retention_array = []
for i in range(len(months)-1):
    retention_data = {}
    selected_month = months[i+1]
    prev_month = months[i]
    retention_data['InvoiceYearMonth'] = int(selected_month)
    retention_data['TotalUserCount'] = tx_retention[selected_month].sum()
    retention_data['RetainedUserCount'] = tx_retention[(tx_retention[selected_month]>0) & (tx_retention[prev_month]>0)][selected_month].sum()
    retention_array.append(retention_data)

In [67]:
#convert the array to dataframe and calculate Retention Rate
tx_retention = pd.DataFrame(retention_array)
tx_retention['RetentionRate'] = tx_retention['RetainedUserCount']/tx_retention['TotalUserCount']
tx_retention 

InvoiceYearMonth  TotalUserCount  RetainedUserCount  RetentionRate
0             201102             714                263       0.368347
1             201103             923                305       0.330444
2             201104             817                310       0.379437
3             201105             985                369       0.374619
4             201106             943                417       0.442206
5             201107             899                379       0.421580
6             201108             867                391       0.450980
7             201109            1177                417       0.354291
8             201110            1285                502       0.390661
9             201111            1548                616       0.397933
10            201112             617                402       0.651540

In [65]:
#plot the retention rate graph
plot_data = [
    go.Scatter(
        x=tx_retention.query("InvoiceYearMonth<201112")['InvoiceYearMonth'],
        y=tx_retention.query("InvoiceYearMonth<201112")['RetentionRate'],
        name="organic"
    )
    
]

plot_layout = go.Layout(
        xaxis={"type": "category"},
        title='Monthly Retention Rate'
    )
fig = go.Figure(data=plot_data, layout=plot_layout)
fig.show()

Monthly Retention Rate significantly jumped from June to August and went back to previous levels afterwards.

### Cohort Based Retention Rate

Retentation rate can be calculated using cohorts. Cohorts are determined as first purchase year-month of the customers
It involves measuring what percentage of the customers retained after their first purchase in each month.


In [68]:
#create our retention table again with crosstab() and add firs purchase year month view
tx_retention = pd.crosstab(tx_user_purchase['CustomerID'], tx_user_purchase['InvoiceYearMonth']).reset_index()
tx_retention = pd.merge(tx_retention,tx_min_purchase[['CustomerID','MinPurchaseYearMonth']],on='CustomerID')
new_column_names = [ 'm_' + str(column) for column in tx_retention.columns[:-1]]
new_column_names.append('MinPurchaseYearMonth')
tx_retention.columns = new_column_names

In [69]:
#create the array of Retained users for each cohort monthly
retention_array = []
for i in range(len(months)):
    retention_data = {}
    selected_month = months[i]
    prev_months = months[:i]
    next_months = months[i+1:]
    for prev_month in prev_months:
        retention_data[prev_month] = np.nan
        
    total_user_count = tx_retention[tx_retention.MinPurchaseYearMonth ==  selected_month].MinPurchaseYearMonth.count()
    retention_data['TotalUserCount'] = total_user_count
    retention_data[selected_month] = 1 
    
    query = "MinPurchaseYearMonth == {}".format(selected_month)
    

    for next_month in next_months:
        new_query = query + " and {} > 0".format(str('m_' + str(next_month)))
        retention_data[next_month] = np.round(tx_retention.query(new_query)['m_' + str(next_month)].sum()/total_user_count,2)
    retention_array.append(retention_data)
    
tx_retention = pd.DataFrame(retention_array)
tx_retention.index = months

#showing new cohort based retention table
tx_retention

TotalUserCount  201101  201102  201103  201104  201105  \
InvoiceYearMonth                                                           
201101                       362     1.0    0.23    0.28    0.25    0.34   
201102                       339     NaN    1.00    0.25    0.19    0.28   
201103                       408     NaN     NaN    1.00    0.19    0.26   
201104                       276     NaN     NaN     NaN    1.00    0.22   
201105                       252     NaN     NaN     NaN     NaN    1.00   
201106                       207     NaN     NaN     NaN     NaN     NaN   
201107                       172     NaN     NaN     NaN     NaN     NaN   
201108                       140     NaN     NaN     NaN     NaN     NaN   
201109                       275     NaN     NaN     NaN     NaN     NaN   
201110                       318     NaN     NaN     NaN     NaN     NaN   
201111                       296     NaN     NaN     NaN     NaN     NaN   
201112                        34     NaN     NaN     NaN     NaN     NaN   

                  201106  201107  201108  201109  201110  201111  201112  
InvoiceYearMonth                                                          
201101              0.29    0.26    0.26    0.31    0.35    0.36    0.15  
201102              0.28    0.25    0.26    0.28    0.28    0.31    0.10  
201103              0.22    0.23    0.17    0.26    0.24    0.29    0.09  
201104              0.22    0.22    0.21    0.23    0.23    0.26    0.08  
201105              0.23    0.17    0.17    0.21    0.24    0.27    0.10  
201106              1.00    0.21    0.16    0.25    0.26    0.32    0.10  
201107               NaN    1.00    0.20    0.19    0.23    0.28    0.11  
201108               NaN     NaN    1.00    0.26    0.23    0.26    0.14  
201109               NaN     NaN     NaN    1.00    0.29    0.33    0.12  
201110               NaN     NaN     NaN     NaN    1.00    0.27    0.13  
201111               NaN     NaN     NaN     NaN     NaN    1.00    0.14  
201112               NaN     NaN     NaN     NaN     NaN     NaN    1.00